### PR TITLE
BUGFIX and improvements in tree_data and tree_data_helpers

### PR DESCRIPTION
**Pull request by lePici**

### CODE REVIEW

File: src/common.h [hunk 1]
```diff
@@ -369,6 +369,24 @@ char *ly_strnchr(const char *s, int c, size_t len);
  */
 int ly_strncmp(const char *refstr, const char *str, size_t str_len);
 
+/**
+ * @brief Get xml namespace prefix or just some substring from @p str.
```
Comment (michalvasko) [inline src/common.h:373]: This is not specific to XML, you can call it `YANG identifier prefix`, I guess.

File: src/common.h [hunk 2]
```diff
@@ -369,6 +369,24 @@ char *ly_strnchr(const char *s, int c, size_t len);
...
+ *
+ * The function looking for possible xml namespace prefix at the beginning of the @p str_begin and return it.
+ * If there is some string at the @p str_begin and then follows the prefix, then the string is returned.
+ * If @p str_begin does not contain any prefix, then the entire @p str_begin is returned.
+ * The @p rest can be used as the @p str_begin in the next iteration.
+ *
+ * @param[in] str_begin Begin of the input string.
+ * @param[in] str_end End of the input string. If set to NULL then the @p str_begin must be NULL-terminated string.
+ * @param[out] substr_type Type of substring found. Set to 1 for the xml namespace prefix, otherwise 0.
+ * @param[out] rest String the remainder part located behind the found substring.
```
Comment (michalvasko) [inline src/common.h:383]: `behind`, are you sure? I would guess it is after.

File: src/common.h [hunk 3]
```diff
@@ -369,6 +369,24 @@ char *ly_strnchr(const char *s, int c, size_t len);
...
+ * @brief Get xml namespace prefix or just some substring from @p str.
+ *
+ * The function looking for possible xml namespace prefix at the beginning of the @p str_begin and return it.
+ * If there is some string at the @p str_begin and then follows the prefix, then the string is returned.
+ * If @p str_begin does not contain any prefix, then the entire @p str_begin is returned.
+ * The @p rest can be used as the @p str_begin in the next iteration.
+ *
+ * @param[in] str_begin Begin of the input string.
+ * @param[in] str_end End of the input string. If set to NULL then the @p str_begin must be NULL-terminated string.
+ * @param[out] substr_type Type of substring found. Set to 1 for the xml namespace prefix, otherwise 0.
```
Comment (michalvasko) [inline src/common.h:382]: Either call it `is_prefix`, for example, or change the type to an enum. Preferably the former.

File: src/common.h [hunk 4]
```diff
@@ -369,6 +369,24 @@ char *ly_strnchr(const char *s, int c, size_t len);
...
+/**
+ * @brief Get xml namespace prefix or just some substring from @p str.
+ *
+ * The function looking for possible xml namespace prefix at the beginning of the @p str_begin and return it.
+ * If there is some string at the @p str_begin and then follows the prefix, then the string is returned.
+ * If @p str_begin does not contain any prefix, then the entire @p str_begin is returned.
+ * The @p rest can be used as the @p str_begin in the next iteration.
+ *
+ * @param[in] str_begin Begin of the input string.
+ * @param[in] str_end End of the input string. If set to NULL then the @p str_begin must be NULL-terminated string.
```
Comment (michalvasko) [inline src/common.h:381]: This information would be more convenient as the string length instead, don't you agree?

File: src/common.h [hunk 5]
```diff
@@ -369,6 +369,24 @@ char *ly_strnchr(const char *s, int c, size_t len);
...
+ * @param[out] substr_type Type of substring found. Set to 1 for the xml namespace prefix, otherwise 0.
+ * @param[out] rest String the remainder part located behind the found substring.
+ * If the @p substr_type is set to 1 then the colon character is skipped.
+ * If no string remains, it is set to NULL.
+ * @return A number of bytes of substring starting at position @p str_begin.
```
Comment (michalvasko) [inline src/common.h:386]: I would reword this to `Length of found substring starting at @p str_begin.`, for example.

**Review by michalvasko — COMMENTED** [review]
Some tweaks needed but the full functionality is there.

File: src/common.h (hunk 5 above)
Comment (lePici) [reply src/common.h:386]: I think the mention of bytes is more accurate, because the `ly_getutf8` function can have up to 4 bytes for one character.

Comment (michalvasko) [reply src/common.h:386]: Fair enough, then make it `Number of bytes of found substring starting at @p str_begin.`. Generally, when documenting this way with doxygen, skip articles at the beginning, that is the custom.

File: src/common.h [hunk 6]
```diff
@@ -369,6 +369,24 @@ char *ly_strnchr(const char *s, int c, size_t len);
...
+ *
+ * The function looking for possible YANG identifier prefix at the beginning of the @p str_begin and return it.
+ * If there is some string at the @p str_begin and then follows the prefix, then the string is returned.
+ * If @p str_begin does not contain any prefix, then the entire @p str_begin is returned.
+ * The @p rest can be used as the @p str_begin in the next iteration.
+ *
+ * @param[in] str_begin Begin of the input string.
+ * @param[in] str_end Length of the @p str_begin. If set to NULL then the @p str_begin must be NULL-terminated string.
+ * @param[out] is_prefix Type of substring found. Set to 1 for prefix, otherwise 0.
+ * @param[out] rest String the remainder part located after the found substring.
```
Comment (michalvasko) [inline src/common.h:383]: I would rename to `next` to be consistent with all similar iterator functions, even with the function name. The comment could then be improved to something like `Rest of the string after the found substring.`.

File: src/tree_data_helpers.c [hunk 7]
```diff
@@ -568,38 +570,26 @@ ly_store_prefix_data(const struct ly_ctx *ctx, const char *value, size_t value_l
             prefixes = *prefix_data_p;
         }
 
+        value_end = value + value_len;
```
Comment (michalvasko) [inline src/tree_data_helpers.c:573]: I would switch this line and the next one, to signify that this line is directly connected to the loop and functionality in it, it is not separate.

File: src/tree_data_helpers.c [hunk 8]
```diff
@@ -616,39 +606,27 @@ ly_store_prefix_data(const struct ly_ctx *ctx, const char *value, size_t value_l
             ns_list = *prefix_data_p;
         }
 
+        value_end = value + value_len;
```
Comment (michalvasko) [inline src/tree_data_helpers.c:609]: Same as the previous comment.

File: src/common.h [hunk 9]
```diff
@@ -369,6 +369,24 @@ char *ly_strnchr(const char *s, int c, size_t len);
  */
 int ly_strncmp(const char *refstr, const char *str, size_t str_len);
 
+/**
+ * @brief Get YANG identifier prefix or just some substring from @p str.
```
[the remark this answers is not in the document]
Comment (lePici) [reply src/common.h]: ok

File: src/common.h [hunk 10]
```diff
@@ -369,6 +369,24 @@ char *ly_strnchr(const char *s, int c, size_t len);
...
+ *
+ * The function looking for possible YANG identifier prefix at the beginning of the @p str_begin and return it.
```
[the remark this answers is not in the document]
Comment (lePici) [reply src/common.h]: ok

File: src/common.h [hunk 11]
```diff
@@ -369,6 +369,24 @@ char *ly_strnchr(const char *s, int c, size_t len);
...
+ * If there is some string at the @p str_begin and then follows the prefix, then the string is returned.
+ * If @p str_begin does not contain any prefix, then the entire @p str_begin is returned.
+ * The @p rest can be used as the @p str_begin in the next iteration.
+ *
+ * @param[in] str_begin Begin of the input string.
+ * @param[in] str_end Length of the @p str_begin. If set to NULL then the @p str_begin must be NULL-terminated string.
+ * @param[out] is_prefix Type of substring found. Set to 1 for prefix, otherwise 0.
+ * @param[out] rest String the remainder part located after the found substring.
+ * If the @p substr_type is set to 1 then the colon character is skipped.
```
[the remark this answers is not in the document]
Comment (lePici) [reply src/common.h]: ok

File: src/common.h (hunk 6 above)
Comment (lePici) [reply src/common.h]: ok

File: src/tree_data_helpers.c (hunk 7 above)
Comment (lePici) [reply src/tree_data_helpers.c]: ok

File: src/tree_data_helpers.c (hunk 8 above)
Comment (lePici) [reply src/tree_data_helpers.c]: ok

File: src/common.h [hunk 12]
```diff
@@ -369,6 +369,24 @@ char *ly_strnchr(const char *s, int c, size_t len);
...
+ *
+ * The function looks for possible prefix at the beginning of the @p str_begin and returns it.
+ * If there is some string at the @p str_begin and then follows the prefix, then the string is returned.
+ * If @p str_begin does not contain any prefix, then the entire @p str_begin is returned.
+ * The @p next can be used as the @p str_begin in the next iteration.
+ *
+ * @param[in] str_begin Begin of the input string.
+ * @param[in] str_end Length of the @p str_begin. If set to NULL then the @p str_begin must be NULL-terminated string.
+ * @param[out] is_prefix Type of substring found. Set to 1 for prefix, otherwise 0.
+ * @param[out] next Rest of the string after the found substring.
```
Comment (michalvasko) [inline src/common.h:383]: Just to be consistent with your other parameter names, maybe call this `str_next`.

Comment (lePici) [reply src/common.h]: ok

File: src/tree_data.c [hunk 13]
```diff
@@ -2796,25 +2879,33 @@ lyd_compare_single(const struct lyd_node *node1, const struct lyd_node *node2, u
         }
     }
 
-    if ((LYD_CTX(node1) != LYD_CTX(node2)) || (node1->schema != node2->schema)) {
+    if (node1->hash != node2->hash) {
         return LY_ENOT;
     }
+    /* equal hashes do not mean equal nodes, they can be just in collision (or both be 0) so the nodes must be checked explicitly */
```
Comment (michalvasko) [inline src/tree_data.c:2885]: I know you did not change this, but I do not think the hashes can be `0` so I would remove that.

File: src/tree_data.c [hunk 14]
```diff
@@ -2847,7 +2938,16 @@ lyd_compare_single(const struct lyd_node *node1, const struct lyd_node *node2, u
...
+
+            if (LYD_CTX(node1) == LYD_CTX(node2)) {
+                return term1->value.realtype->plugin->compare(&term1->value, &term2->value);
+            } else {
+                if (strcmp(LYD_CANON_VALUE(node1), LYD_CANON_VALUE(node2))) {
+                    return LY_ENOT;
+                } else {
+                    return LY_SUCCESS;
+                }
+            }
```
Comment (michalvasko) [inline src/tree_data.c:2950]: It took me a few seconds to notice that `break` is not missing because a `return` must be called but you can save these seconds for any next person by refactoring this code a bit. Something like this would be sufficient but it is up to you to improve it.
```C
            if (LYD_CTX(node1) == LYD_CTX(node2)) {
                return term1->value.realtype->plugin->compare(&term1->value, &term2->value);
            }

            /* different context */
            if (strcmp(LYD_CANON_VALUE(node1), LYD_CANON_VALUE(node2))) {
                return LY_ENOT;
            } else {
                return LY_SUCCESS;
            }
```
One `if-else` is fine, but not nested.

File: src/tree_data_helpers.c [hunk 15]
```diff
@@ -567,39 +569,28 @@ ly_store_prefix_data(const struct ly_ctx *ctx, const char *value, size_t value_l
             assert(*format_p == LY_PREF_SCHEMA_RESOLVED);
             prefixes = *prefix_data_p;
         }
+        //TODO remove
```
Comment (michalvasko) [inline src/tree_data_helpers.c:572]: I suppose all of these should be removed...

File: src/xml.c [hunk 16]
```diff
@@ -1100,38 +1104,75 @@ lyxml_value_compare(const struct ly_ctx *ctx, const char *value1, void *val_pref
...
+            mod2 = ly_resolve_prefix(ctx2, value2_iter, value2_len, LY_PREF_XML, val_prefix_data2);
+        }
+        if (!mod1 || !mod2) {
+            /* not a prefix or maps to different namespaces */
+            ret = LY_ENOT;
+            break;
+        }
+
+        if (mod1->ctx == mod2->ctx) {
+            if ((mod1->name != mod2->name) || (mod1->revision != mod2->revision)) {
```
Comment (michalvasko) [inline src/xml.c:1150]: I would like a comment `/* same contexts */` above this line.

File: src/xml.c [hunk 17]
```diff
@@ -1100,38 +1104,75 @@ lyxml_value_compare(const struct ly_ctx *ctx, const char *value1, void *val_pref
...
+                ret = LY_ENOT;
+                break;
+            }
+        } else {
+            if (strcmp(mod1->name, mod2->name)) {
```
Comment (michalvasko) [inline src/xml.c:1155]: And `/* different contexts */` above this.

File: src/tree_data.c [hunk 18]
```diff
@@ -2779,6 +2779,82 @@ lyd_target(const struct ly_path *path, const struct lyd_node *tree)
...
+    }
+
+    if (parent1 || parent2) {
+        return 1;
+    }
+
+    return 0;
+}
+
+
```
Comment (michalvasko) [inline src/tree_data.c:2857]: Extra line.

File: src/tree_data.c [hunk 19]
```diff
@@ -2788,6 +2864,13 @@ lyd_compare_single(const struct lyd_node *node1, const struct lyd_node *node2, u
     struct lyd_node_opaq *opaq1, *opaq2;
     size_t len1, len2;
 
+/* Internal flag for lyd_compare_single function.
+ * This optimization flag will be set in the options variable
+ * if the lyd_compare_parental_schemas function is called.
+ * As a result, the lyd_compare_parental_schemas is performed only once.
+ */
+#define LYD_PARENTAL_SCHEMAS_COMPARED 0x80000000
```
Comment (michalvasko) [inline src/tree_data.c:2872]: I do not like mixing API and internal flags in one variable but I suppose I could let this one slide.

File: src/tree_data.c [hunk 20]
```diff
@@ -2779,6 +2779,82 @@ lyd_target(const struct ly_path *path, const struct lyd_node *tree)
...
+ */
+static ly_bool
+lyd_compare_schemas(const struct lysc_node *schema1, const struct lysc_node *schema2)
+{
+    if (schema1 == schema2) {
+        return 0;
+    } else if (!schema1 || !schema2) {
+        return 1;
+    }
+
```
Comment (michalvasko) [inline src/tree_data.c:2797]: Here I would expect this code, with some comment even to explain it.
```C
if (schema1->module->ctx == schema2->module->ctx) {
    return 1;
}
```
Keep the same-context efficiency.

File: src/tree_data.c [hunk 21]
```diff
@@ -2779,6 +2779,82 @@ lyd_target(const struct ly_path *path, const struct lyd_node *tree)
...
+ * @brief Compare all parental schemas of two data nodes.
+ *
+ * @param schema1 of first node.
+ * @param schema2 of second node.
+ * @return 0 if the all parental schemas are equivalent otherwise 1.
+ */
+static ly_bool
+lyd_compare_parental_schemas(const struct lyd_node *node1, const struct lyd_node *node2)
+{
+    const struct lyd_node_inner *parent1, *parent2;
```
Comment (michalvasko) [inline src/tree_data.c:2836]: Just a hint, to avoid having to use `struct lyd_node_inner *` you can use the function `lyd_parent()`, it returns `struct lyd_node *`.

File: src/tree_data.c [hunk 22]
```diff
@@ -2779,6 +2779,82 @@ lyd_target(const struct ly_path *path, const struct lyd_node *tree)
...
+
+    if (node1->schema == node2->schema) {
```
Comment (michalvasko) [inline src/tree_data.c:2838]: You already did this comparison, is there any purpose to doing it again? You can put an assert, if you to make sure it works the way it should.

File: src/tree_data.c [hunk 23]
```diff
@@ -2779,6 +2779,82 @@ lyd_target(const struct ly_path *path, const struct lyd_node *tree)
...
+}
+
+/**
+ * @brief Compare all parental schemas of two data nodes.
+ *
+ * @param schema1 of first node.
+ * @param schema2 of second node.
+ * @return 0 if the all parental schemas are equivalent otherwise 1.
+ */
+static ly_bool
```
Comment (michalvasko) [inline src/tree_data.c:2833]: You are returning a boolean so I would expect `1` to be equal, `0` to be different but it inverted. Also, the name `compare` would imply return values just like `strcmp` but they are not, so this prototype is confusing. My idea would be to keep the boolean with inverted values and rename the function to `lyd_compare_schema_parents_equal`. No need for `schemas` (you cannot compare a single schema), `lyd_compare` prefix means it is specific to `lyd_compare` functions (not sure it is, once it is used somewhere else, it should be renamed), and `equal` means it is a predicate (boolean return).

File: src/tree_data.c [hunk 24]
```diff
@@ -2779,6 +2779,82 @@ lyd_target(const struct ly_path *path, const struct lyd_node *tree)
...
 
+/**
+ * @brief Compare schemas of two Compiled YANG data nodes.
+ *
+ * @param schema1 of first node.
+ * @param schema2 of second node.
+ * @return 0 if the schemas are equivalent otherwise 1.
+ */
+static ly_bool
+lyd_compare_schemas(const struct lysc_node *schema1, const struct lysc_node *schema2)
```
Comment (michalvasko) [inline src/tree_data.c:2790]: Rename to `lyd_compare_schema_equal`, same rationale as for the parent variant.

File: src/tree_data.c (hunk 19 above)
Comment (lePici) [reply src/tree_data.c:2872]: So no changes are required?

File: src/tree_data.c (hunk 20 above)
Comment (lePici) [reply src/tree_data.c:2797]: why is this necessary? I thought that when the pointers `schema1 == schema2` are equal, they are in the same contexts. Isn't that so?

File: src/tree_data.c (hunk 22 above)
Comment (lePici) [reply src/tree_data.c:2838]: assert won't help here. The purpose of this condition is optimization. Thanks to it, it does not have to iterate through the `for `cycle below. I assume that if the schemes are equal, all parent nodes must be equal.

File: src/tree_data.c (hunk 24 above)
Comment (lePici) [reply src/tree_data.c:2790]: ok

File: src/tree_data.c (hunk 23 above)
Comment (lePici) [reply src/tree_data.c:2833]: ok

File: src/tree_data.c (hunk 21 above)
Comment (lePici) [reply src/tree_data.c:2836]: ok

File: src/tree_data.c (hunk 18 above)
Comment (lePici) [reply src/tree_data.c:2857]: ok

File: src/xml.c (hunk 17 above)
Comment (lePici) [reply src/xml.c:1155]: ok

File: src/xml.c (hunk 16 above)
Comment (lePici) [reply src/xml.c:1150]: ok

File: src/tree_data_helpers.c (hunk 15 above)
Comment (lePici) [reply src/tree_data_helpers.c:572]: yes

File: src/tree_data.c (hunk 20 above)
Comment (michalvasko) [reply src/tree_data.c:2797]: But what about the situation when they are in the same context but `schema1 != schema2`? Right now the whole comparison for different context is performed, which is completely useless. What you should generally do is strictly distinguish in every comparison whether the contexts are the same or not, it save you these kinds of problems.

File: src/tree_data.c (hunk 14 above)
Comment (lePici) [reply src/tree_data.c:2950]: ok

File: src/tree_data.c (hunk 13 above)
Comment (lePici) [reply src/tree_data.c:2885]: ok

File: src/tree_data.c (hunk 22 above)
Comment (michalvasko) [reply src/tree_data.c:2838]: Right, I have forgotten a comment I wanted to put here, again distinguishing between the same context and a different one. You do not do that at all and only comparing the schemas for equality is not enough, as I have mentioned.

File: src/tree_data.c (hunk 19 above)
Comment (michalvasko) [reply src/tree_data.c:2872]: I guess, I will let @rkrejci decide.

File: src/tree_data.c (hunk 20 above)
Comment (lePici) [reply src/tree_data.c:2797]: ok, I get it

File: src/tree_data.c [hunk 25]
```diff
@@ -2779,6 +2779,83 @@ lyd_target(const struct ly_path *path, const struct lyd_node *tree)
...
+        return 0;
+    }
+
+    assert(schema1->module->ctx != schema2->module->ctx);
+
+    if (schema1->nodetype != schema2->nodetype) {
+        return 0;
+    }
+
+    if (schema1->module->latest_revision != schema2->module->latest_revision) {
```
Comment (michalvasko) [inline src/tree_data.c:2804]: Ah, another thing I forgot about before, but no harm done in this case. This is redundant and actually wrong, ignore those `latest_revision` flags, they have no relevance for this purpose.

File: src/common.h [hunk 26]
```diff
@@ -369,6 +369,24 @@ char *ly_strnchr(const char *s, int c, size_t len);
...
+ *
+ * @param[in] str_begin Begin of the input string.
+ * @param[in] str_end Length of the @p str_begin. If set to NULL then the @p str_begin must be NULL-terminated string.
+ * @param[out] is_prefix Type of substring found. Set to 1 for prefix, otherwise 0.
+ * @param[out] str_next Rest of the string after the found substring.
+ * If the @p is_prefix is set to 1 then the colon character is skipped.
+ * If no string remains, it is set to NULL.
+ * @return Number of bytes (length) of the found substring starting at @p str_begin.
+ */
+size_t ly_string_next(const char *str_begin, const char *str_end, int *is_prefix, const char **str_next);
```
Comment (michalvasko) [inline src/common.h:388]: I have not noticed before, why is `is_prefix` `int`? It should definitely be a `ly_bool`.

File: src/xml.c [hunk 27]
```diff
@@ -1100,38 +1104,77 @@ lyxml_value_compare(const struct ly_ctx *ctx, const char *value1, void *val_pref
...
+            value1_iter = value1_next, value2_iter = value2_next) {
+        value1_len = ly_string_next(value1_iter, NULL, &is_prefix1, &value1_next);
+        value2_len = ly_string_next(value2_iter, NULL, &is_prefix2, &value2_next);
+
+        if (is_prefix1 != is_prefix2) {
+            ret = LY_ENOT;
+            break;
+        }
+
+        if (is_prefix1 != 1) {
```
Comment (michalvasko) [inline src/xml.c:1123]: Connected to a previous comment, it is a boolean value (even based on the name), so treat it that way.

**Review by michalvasko — COMMENTED** [review]
Just a side note, be careful when using `size_t` type because it is not possible (or insanely complex) to print such a value in a portable way. It does not matter in this case but the general rule is sticking to basic `stdint` types.

File: src/tree_data.c (hunk 25 above)
Comment (lePici) [reply src/tree_data.c:2804]: It is interesting that it has to be deleted. If I delete it, it does not pass tests in sysrepo. Are you sure about that?

Comment (michalvasko) [reply src/tree_data.c:2804]: Yes, pretty sure. Please, never adjust the code for the tests without understanding the specific test-case. So, when you understand what exactly is going on (what I can help with), let me know and we can discuss what to do.

File: src/xml.h [hunk 28]
```diff
@@ -164,16 +164,17 @@ void lyxml_ctx_free(struct lyxml_ctx *xmlctx);
 /**
  * @brief Compare values and their prefix mappings.
  *
- * @param[in] ctx libyang context.
+ * @param[in] ctx1 libyang context of First value.
```
Comment (rkrejci) [inline src/xml.h:167]: refer to the parameter: `@p value1`

**Review by rkrejci — CHANGES_REQUESTED** [review]

File: src/xml.h (hunk 28 above)
Comment (michalvasko) [reply src/xml.h:167]: It was my initial idea as well but I let it be since `value1` does not have a context, strictly speaking. So the description will have to be a bit more verbose, something like `ctx1 libyang context for resolving prefixes in @p value1`.

File: src/common.h (hunk 26 above)
[the remark this answers is not in the document]
Comment (michalvasko) [reply src/common.h:388]: Not sure if @lePici is able to come up with something accurate enough, I think we should help. The idea is that this function can take any string in a YANG module (hence called YANG string) such as an extension name but even an instance-identifier default value and iteratively return for the caller all possible prefixes (substrings that could be interpreted as a prefix if the whole string conforms to some subset of XPath). The use-case is quite simple, I think, but accurately describing this function is not. Can you @rkrejci suggest some specific improvements based on what I wrote?

Now that I think about, maybe I was forcing this function to be more generic than is needed because for cases such as the extension instance name, this function is not (and will likely never be) used because of strict syntax restrictions. Meaning it will be used only for YANG data values (and YANG schema default values). So maybe `YANG string` can be switched for `YANG value` and the function renamed to `ly_value_next`.

Comment (lePici) [reply src/common.h:388]: ok

File: src/xml.c (hunk 27 above)
Comment (lePici) [reply src/xml.c:1123]: ok

File: src/common.h (hunk 26 above)
[the remark this answers is not in the document]
Comment (lePici) [reply src/common.h:388]: I will use the names `ly_value_prefix_next` and `YANG value` for now.

File: src/xml.h (hunk 28 above)
Comment (lePici) [reply src/xml.h:167]: ok

File: src/common.c [hunk 29]
```diff
@@ -66,6 +66,73 @@ ly_strncmp(const char *refstr, const char *str, size_t str_len)
     }
 }
 
+uint64_t
```
Comment (michalvasko) [inline src/common.c:69]: I mean, it probably does not matter but this is redundant, `uint32_t` is more than enough.

File: src/tree_data.c (hunk 18 above)
Comment (michalvasko) [reply src/tree_data.c:2855]: Does not seem fixed.

File: src/common.h [hunk 30]
```diff
@@ -369,6 +369,36 @@ char *ly_strnchr(const char *s, int c, size_t len);
...
+ * Examples of inputs and outputs are shown in the table below.
+ * Output string @p str_next is used in the next iteration as input parameter @p str_begin.
+   @verbatim
+   |            INPUT              |                         OUTPUT                        |
+   |                               |   iteration 1    |   iteration 2    |   iteration 3   |
+   |------------------------------ |------------------|------------------|-----------------|
+   | /namespace_prefix:some_string | /                | namespace_prefix | some_string     |
+   | namespace_prefix:some_string  | namespace_prefix | some_string      | NULL            |
+   | /some_string                  | /some_string     | NULL             | NULL            |
+   @endverbatim
```
Comment (rkrejci) [inline src/common.h:389]: I miss the example regarding the case 2 described above, probably this (not sure if it is correct, especially regarding the semicolon character):
```
| string1;string2     | string1;  | string2 | NULL
```

**Review by rkrejci — CHANGES_REQUESTED** [review]

File: src/common.h (hunk 30 above)
Comment (michalvasko) [reply src/common.h:389]: Semicolon is not treated specially in any way, there is nothing in the description hinting this, right? It should simply be
```
| string1;string2     | string1;string2 | NULL
```
Right @lePici ?

Comment (lePici) [reply src/common.h:389]: Yes only the colon is taken into account. I will add information about the colon to the documentation. But if there is a use-case with a semicolon somewhere, I can extend the function specification.

Comment (michalvasko) [reply src/common.h:389]: No, to my knowledge it would be redundant to do anything with a semicolon so I do not know why @rkrejci asked about it.

Comment (rkrejci) [reply src/common.h:389]: ok, what about `string1;prefix:something`? That's the case that matters, right? My point is that point 2 in the description talks about some substring before the prefix, but it is not specified how the substring and prefix are delimited.

Comment (rkrejci) [reply src/common.h:389]: and it's not only semicolon, it is recognized using `is_xmlqnamechar()`

Comment (michalvasko) [reply src/common.h:389]: Oh, okay, this information can be added, I suppose. Can you suggest something concise? The description was a bit general because I do not think you need to know the exact delimiting characters to be able to use it, you just know that you have a YANG value and want to get the prefixes from it, so you will use this function. Also, for this to work perfectly the input format should probably matter as in JSON the prefixes can only be YANG module names meaning a subset of the characters allowed for XML. But even if something else is returned for JSON, it will never match a module name, so no harm done.

Comment (lePici) [reply src/common.h:389]: ok, I'll add this comment to the description.

**Review by michalvasko — APPROVED** [review]

**Review by rkrejci — APPROVED** [review]
